### ISSUE TITLE
what to do about iTLB misses

### ISSUE THREAD

**Issue by akuzm** [ClickHouse/ClickHouse#14519]
ClickHouse binary is huge, on the order of gigabyte. We're seeing some instability in performance tests that is related to this -- e.g. `joins_in_memory_pmj` is notorious. Recently, after merging bigint support, the binary size increased and now we have 400 unstable queries instead of 200.

The thing we should (probably) do is optimize the binary size, but it's not the simplest.

What else we can try:
- Use huge pages for executable, with help of [libhugetlbfs](https://github.com/libhugetlbfs/libhugetlbfs). It requires relinking the application with `-z max-page-size=2M` and linking it with `libhugetlbfs`. The library will then allocate the huge pages and copy the executable there. Their [`HOWTO`](https://github.com/libhugetlbfs/libhugetlbfs/blob/master/HOWTO) is pretty helpful on this.
An obstacle is that we don't have huge pages enabled in Sandbox environment where we run the performance tests.
- Link with PGO. Collect PGO info on master, running the performance test queries.
- Link with [Google Propeller](https://github.com/google/llvm-propeller) which does even more PGO. There is also Facebook BOLT but it uses more memory and is not integrated with clang.

**Comment by alexey-milovidov** [comment]
> The library will then allocate the huge pages and copy the executable there.

We can do it by simply call of `madvise` - it will work if transparent huge pages set to "on demand" which is the default.

**Comment by alexey-milovidov** [comment]
> An obstacle is that we don't have huge pages enabled

```
$ cat /sys/kernel/mm/transparent_hugepage/enabled
always [madvise] never
```

it should be set to `madvise`.

**Comment by akuzm** [comment]
> > An obstacle is that we don't have huge pages enabled
> 
> ```
> $ cat /sys/kernel/mm/transparent_hugepage/enabled
> always [madvise] never
> ```
> 
> it should be set to `madvise`.

This one is set, but the huge page pool is empty:
```
$ cat /proc/meminfo
....
HugePages_Total:       0
....
```
Not sure I can configure this in Sandbox.


> We can do it by simply call of madvise - it will work if transparent huge pages set to "on demand" which is the default.

For mmapped memory, sure, but would it work for the binary? I'm not sure, couldn' find any reference for this.

**Comment by alexey-milovidov** [comment]
> huge page pool is empty

Static huge pages are unneeded for using `madvise`, it will work nevertheless.

**Comment by alexey-milovidov** [comment]
> but would it work for the binary? I'm not sure, couldn' find any reference for this.

Me neither. Probably not. But we can use the trick with copy.

**Comment by alexey-milovidov** [comment]
Fixed in #14685
